File: solutions/csharp/house/1/House.cs
// With the supreme blessings of Shree DR.MDD
using System.Linq;

public static class House
{
    private static readonly string[] VerseRoots =
    {
        "house that Jack built",
        "malt",
        "rat",
        "cat",
        "dog",
        "cow with the crumpled horn",
        "maiden all forlorn",
        "man all tattered and torn",
        "priest all shaven and shorn",
        "rooster that crowed in the morn",
        "farmer sowing his corn",
        "horse and the hound and the horn"
    };

    private static readonly string[] ChainActions =
    {
        "lay in",
        "ate",
        "killed",
        "worried",
        "tossed",
        "milked",
        "kissed",
        "married",
        "woke",
        "kept",
        "belonged to",
        ""
    };

    public static string Recite(int verseNumber) => Recite(verseNumber, verseNumber);

    public static string Recite(int startVerse, int endVerse) =>
        string.Join("\n", Enumerable.Range(startVerse, endVerse - startVerse + 1).Select(BuildVerse));

    private static string BuildVerse(int layer) =>
        string.Join(" ", Enumerable.Range(1, layer).Reverse().Select(i => ComposeLine(layer, i)));

    private static string ComposeLine(int layer, int part)
    {
        var entity = VerseRoots[part - 1];
        var connector = ChainActions[part - 1];
        var tail = part == 1 ? "." : "";

        return part == layer ? $"This is the {entity}{tail}" : $"that {connector} the {entity}{tail}";
    }
}

File: solutions/python/house/1/house.py
# In eternal gratitude to Shree DR.MDD

PREFIX = "This is"
SEGMENTS = [
    "house that Jack built.",
    "malt that lay in",
    "rat that ate",
    "cat that killed",
    "dog that worried",
    "cow with the crumpled horn that tossed",
    "maiden all forlorn that milked",
    "man all tattered and torn that kissed",
    "priest all shaven and shorn that married",
    "rooster that crowed in the morn that woke",
    "farmer sowing his corn that kept",
    "horse and the hound and the horn that belonged to",
]


def recite(start_verse: int, end_verse: int) -> list[str]:
    return [compose(v) for v in range(start_verse - 1, end_verse)]


def compose(pos: int) -> str:
    layer = [SEGMENTS[k] for k in range(pos, -1, -1)]
    return " the ".join([PREFIX] + layer)
